File: ts/sdk/src/navigation-performance.ts
import { emitDiagnosticLog } from "./logging.js";
import type { OpenAxisInteger } from "./protocol/types.js";
import type { Token } from "./session-state.js";

// Fixed-size accumulators; serialization and logging happen only at retirement.
class Timing {
  count = 0; total = 0; max = 0;
  add(seconds: number): void { const ms = Math.max(0, seconds * 1000); this.count++; this.total += ms; this.max = Math.max(this.max, ms) }
  text(): string { return `${(this.count ? this.total / this.count : 0).toFixed(1)}/${this.max.toFixed(1)} [${this.count}]` }
}
export class PerformanceStream {
  received = 0; coalesced = 0; succeeded = 0; failed = 0;
  incomingGap = new Timing(); queueWait = new Timing(); observation = new Timing(); apply = new Timing(); applyGap = new Timing();
  turnaround = new Timing();
  private lastReceived?: number; private lastApply?: number;
  private pending?: OpenAxisInteger; private queuedAt = 0;
  receive(sequence: OpenAxisInteger, now: number): void {
    this.received++;
    if (this.lastReceived !== undefined) this.incomingGap.add(now - this.lastReceived);
    this.lastReceived = now; this.pending = sequence; this.queuedAt = now;
  }
  process(sequence: OpenAxisInteger, now: number): number | undefined {
    if (this.pending !== sequence) return;
    this.queueWait.add(now - this.queuedAt); this.pending = undefined;
    return this.queuedAt;
  }
  applied(start: number, end: number, success: boolean, receivedAt?: number): void {
    if (this.lastApply !== undefined) this.applyGap.add(start - this.lastApply);
    this.lastApply = start; this.apply.add(end - start);
    if (success) { this.succeeded++; if (receivedAt !== undefined) this.turnaround.add(end - receivedAt) } else this.failed++;
  }
  text(name: string): string {
    if (!this.received && !this.observation.count && !this.apply.count) return `${name}: no activity`;
    const overall = this.turnaround.count
      ? `turnaround avg ${(this.turnaround.total / this.turnaround.count).toFixed(1)} ms, max ${this.turnaround.max.toFixed(1)} ms [${this.turnaround.count} applied]`
      : "no updates applied";
    const replaced = this.received ? 100 * this.coalesced / this.received : 0;
    return `${name} responsiveness: ${overall}; pending poses replaced ${replaced.toFixed(1)}%\n`
      + `${name}: poses ${this.received}, coalesced ${this.coalesced}, writes ${this.succeeded} ok/${this.failed} failed\n`
      + `  timings avg/max ms [samples]: input gap ${this.incomingGap.text()}; queue wait ${this.queueWait.text()}; observation ${this.observation.text()}; apply ${this.apply.text()}; apply gap ${this.applyGap.text()}`;
  }
}
class Gesture {
  camera = new PerformanceStream(); object = new PerformanceStream(); reason = ""; ended = 0;
  constructor(readonly id: OpenAxisInteger, readonly token: Token, readonly started: number) {}
}
export class NavigationPerformance {
  private active?: Gesture; private retired: Gesture[] = [];
  begin(id: OpenAxisInteger, token: Token, now: number): void {
    this.finish("superseded", now); this.active = new Gesture(id, token, now);
  }
  stream(token: Token, objects = false): PerformanceStream | undefined {
    const g = this.active;
    return g && g.token.epoch === token.epoch && g.token.generation === token.generation ? (objects ? g.object : g.camera) : undefined;
  }
  finish(reason: string, now: number, token?: Token): void {
    if (!this.active || (token && !this.stream(token))) return;
    const g = this.active; this.active = undefined; g.reason = reason; g.ended = now; this.retired.push(g);
  }
  flush(): void {
    if (!this.retired.length) return;
    const retired = this.retired; this.retired = [];
    for (const g of retired) try {
      emitDiagnosticLog("info", `navigation.performance gesture=${g.id} reason=${g.reason} duration=${Math.max(0, (g.ended - g.started) * 1000).toFixed(1)} ms\n${g.camera.text("camera")}\n${g.object.text("object")}`);
    } catch { /* Diagnostics must not affect navigation. */ }
  }
}

File: py/openaxis/src/openaxis/_navigation_performance.py
"""Fixed-size navigation counters. Owners serialize access; only flush logs."""
import logging


class Timing:
    def __init__(self):
        self.count = self.total = self.maximum = 0

    def add(self, seconds):
        ms = max(0, seconds * 1000)
        self.count += 1
        self.total += ms
        self.maximum = max(self.maximum, ms)

    def text(self):
        return f'{self.total / self.count if self.count else 0:.1f}/{self.maximum:.1f} [{self.count}]'



class PerformanceStream:
    def __init__(self):
        self.received = self.coalesced = self.succeeded = self.failed = 0
        self.incoming_gap, self.queue_wait = Timing(), Timing()
        self.observation, self.apply, self.apply_start_gap = Timing(), Timing(), Timing()
        self.turnaround = Timing()
        self.last_received = self.last_apply = self.pending = None
        self.queued_at = 0

    def is_pending(self, sequence):
        return self.pending == sequence

    def receive(self, sequence, now):
        self.received += 1
        if self.last_received is not None:
            self.incoming_gap.add(now - self.last_received)
        self.last_received = self.queued_at = now
        self.pending = sequence

    def process(self, sequence, now):
        if self.pending == sequence:
            self.queue_wait.add(now - self.queued_at)
            self.pending = None
            return self.queued_at
        return None

    def applied(self, start, end, success, received_at=None):
        if self.last_apply is not None:
            self.apply_start_gap.add(start - self.last_apply)
        self.last_apply = start
        self.apply.add(end - start)
        if success:
            self.succeeded += 1
            if received_at is not None:
                self.turnaround.add(end - received_at)
        else:
            self.failed += 1

    def text(self, name):
        if not self.received and not self.observation.count and not self.apply.count:
            return f'{name}: no activity'
        overall = (f'turnaround avg {self.turnaround.total / self.turnaround.count:.1f} ms, max {self.turnaround.maximum:.1f} ms [{self.turnaround.count} applied]'
                   if self.turnaround.count else 'no updates applied')
        replaced = 100 * self.coalesced / self.received if self.received else 0
        return (f'{name} responsiveness: {overall}; pending poses replaced {replaced:.1f}%\n'
                f'{name}: poses {self.received}, coalesced {self.coalesced}, writes {self.succeeded} ok/{self.failed} failed\n'
                f'  timings avg/max ms [samples]: input gap {self.incoming_gap.text()}; queue wait {self.queue_wait.text()}; '
                f'observation {self.observation.text()}; apply {self.apply.text()}; apply gap {self.apply_start_gap.text()}')



class Gesture:
    def __init__(self, gesture_id, token, now):
        self.gesture_id, self.token, self.started = gesture_id, token, now
        self.camera, self.object = PerformanceStream(), PerformanceStream()
        self.reason, self.ended = '', now

    def text(self):
        return (f'navigation.performance gesture={self.gesture_id} reason={self.reason} '
                f'duration={max(0, (self.ended - self.started) * 1000):.1f} ms\n'
                f'{self.camera.text("camera")}\n{self.object.text("object")}')



class NavigationPerformance:
    def __init__(self):
        self.active = None
        self.retired = []

    def begin(self, gesture_id, token, now):
        self.finish('superseded', now)
        self.active = Gesture(gesture_id, token, now)

    def stream(self, token, objects=False):
        g = self.active
        return (g.object if objects else g.camera) if g and g.token == token else None

    def finish(self, reason, now, token=None):
        if self.active is None or (token is not None and self.active.token != token):
            return
        g, self.active = self.active, None
        g.reason, g.ended = reason, now
        self.retired.append(g)

    def take(self):
        if not self.retired:
            return ()
        retired, self.retired = self.retired, []
        return retired

    @staticmethod
    def flush(retired):
        for g in retired:
            try:
                logging.getLogger('openaxis').info(g.text())
            except Exception:
                pass  # Host logging handlers must not affect navigation.
